invictus_dashboard/src/app/dashboard/dashboard-task-edit.component.ts

```diff
@@ -0,0 +1,324 @@
+      id: [this.taskId],
+      title: ['', [Validators.required]],
+      difficulty: [1, [Validators.required, NumberValidators.range(1, 5)]],
+      priority: [taskPriority.standard, [Validators.required]],
+      completed: [false, [Validators.required]],
+
+      description: [''],
+      startDate: [new Date()],
+      dueDate: [this.sprintEndDateString],
+      subTasks: this.fb.array([]),
+    });
+  }
+
+  addSubTask(): void {
+    const numSubTasks = this.subTasks.length;
+    console.log('adding subtask ' + numSubTasks);
+
+    this.subTasks.push(this.buildTaskGroup());
+    const subTaskGroup: FormGroup = (
+      this.taskForm.get('subTasks') as FormArray
+    ).at(numSubTasks) as FormGroup;
+
+    var subTaskId: number = numSubTasks + 1;
+    while (subTaskId >= 1) {
+      subTaskId /= 10;
+    }
+
+    subTaskId += this.taskForm.get('id').value;
+
+    subTaskGroup.get('id').setValue(subTaskId);
+
+    console.log(`sub-task id: ${subTaskGroup.get('id').value}`);
+
+    this.addGroupValueChangeSubscriptions(subTaskGroup, numSubTasks);
+  }
+
+  deleteSubTask(index: number) {
+    const subTask = (this.taskForm.get('subTasks') as FormArray).at(index);
+    subTask.clearValidators();
+    subTask.updateValueAndValidity();
+    this.subTasks.removeAt(index);
+    this.subTasks.markAsDirty();
+  }
+
+  getTask(id: number) {
+    this.dashboardService.getTask(id).subscribe({
+      next: (task: DashboardTask) => this.displayTask(task),
+      error: (err) => (this.apiErrMsg = err),
+    });
+  }
+
+  displayTask(task: DashboardTask) {
+    if (this.taskForm) {
+      this.taskForm.reset();
+    }
+    this.task = task;
+
+    if (this.task.id === 0) {
+      this.pageTitle = 'Add Task';
+    } else {
+      this.pageTitle = 'Edit ' + this.task.title;
+    }
+
+    this.taskForm.patchValue({
+      id: this.task.id,
+      title: this.task.title,
+      difficulty: this.task.difficulty,
+      priority: this.task.priority,
+      completed: this.task.completed,
+
+      description: this.task.description,
+      startDate: this.task.startDate,
+      dueDate: this.task.dueDate,
+    });
+
+    for (var index in this.task.subTasks) {
+      this.addSubTask();
+      var numSubTasks = this.subTasks.length;
+      this.subTasks.at(numSubTasks - 1).patchValue({
+        // id: this.task.id,
+        title: this.task.subTasks[index].title,
+        completed: this.task.subTasks[index].completed,
+
+        description: this.task.subTasks[index].description,
+        startDate: this.task.subTasks[index].startDate,
+        dueDate: this.task.subTasks[index].dueDate,
+      });
+    }
+  }
+
+  onSave() {
+    if (this.taskForm.valid) {
+      if (this.taskForm.dirty) {
+        console.log('Saving...');
+        const t: DashboardTask = { ...this.task, ...this.taskForm.value };
+        console.log(t);
+
+        if (t.id === 0) {
+          console.log(`creating task: ${t.title}`);
+          this.dashboardService.createTask(t).subscribe({
```

Comment:
Any time you subscribe to an observable you have to unsubscribe or tell it when to stop listening for events. Otherwise it will leave a listener open for the lifetime of the module even after you leave the page. There are a ton of ways to unsubscribe. I just put a bunch of comments in another repo about this that you might find helpful.

https://github.com/efoglia/bootcamp-project/pull/4/files#r922450163

You should also catch up with Emily to talk through if if you need to. I know she's going through some Pluralsight videos on these topics specifically.